rest-client/src/main/java/org/sagebionetworks/bridge/rest/UserSessionInfoProvider.java

```diff
@@ -1,76 +1,99 @@
-     *         sign in credentials for a user
-     * @return session info for the user, or null if the return value was not a 200
+     * @return session info for the user
      * @throws IOException problem while signing in
      */
-    public UserSessionInfo retrieveSession(SignIn signIn) throws IOException {
-        UserSessionInfo session = retrieveCachedSession(signIn);
-
+    synchronized UserSessionInfo retrieveSession() throws IOException {
         if (session != null) {
             return session;
         }
-
-        LOG.debug("No session, call intercepted for authentication attempt: " + signIn.getEmail());
-        authenticationApi.signIn(signIn).execute();
-
-        // Calling sign in will cause the session to be saved by the session interceptor. No further
-        // work is required here.
-        return sessionInterceptor.getSession(signIn);
+        // There's no session, so our only option is to sign in again.
+        if (LOG.isDebugEnabled()) {
+            LOG.debug(debugString("null")+ " signing in");
+        }
+        signIn();
+        return session;
     }
-
-    /**
-     * Retrieves a session without making a network call.
-     *
-     * @param signIn
-     *         sign in credentials for a user
-     * @return session info for the user
-     */
-    public UserSessionInfo retrieveCachedSession(SignIn signIn) {
-        if (signIn == null) {
-            return null;
+    
+    synchronized void setSession(UserSessionInfo session) {
+        if (LOG.isDebugEnabled()) {
+            String token = (session==null) ? "null" : session.getSessionToken();
+            LOG.debug(debugString(token));
         }
-
-        return sessionInterceptor.getSession(signIn);
+        this.session = session;
     }
-
-    /**
-     * Removes a session from the cache. This should be called if a session returned by this instance turns out to be
-     * no longer valid.
-     *
-     * @param session the user's session
-     */
-    void purgeCachedSession(UserSessionInfo session) {
-        sessionInterceptor.removeSession(session);
+    
+    synchronized void reauthenticate() throws IOException {
+        if (session == null || session.getReauthToken() == null) {
+            signIn();
+        } else {
+            ReauthenticateRequest request = new ReauthenticateRequest().email(session.getEmail())
+                    .reauthToken(session.getReauthToken()).study(signIn.getStudy());
+            try {
+                UserSessionInfo newSession = authenticationApi.reauthenticate(request).execute().body();
+                setSession(newSession);
+            } catch(ConsentRequiredException e) {
+                setSession(e.getSession());
+            } catch(Exception e) {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("reauth failed, signing in again");
+                }
+                signIn();
+            }
+        }
+    }
+    
+    private void signIn() throws IOException {
```

Comment:
We're eventually going to need to look at the credentials, and trigger the appropriate authentication pathway (may not be an email/password sign in).

Reply:
Yea, I could see a common app pattern would be prompting for email/password for initial sign in, storing only the (encrypted) reauthToken on the device, and prompting again for email/password again if reauth fails.

This would fit UXs I've seen with pin/fingerprint secured reauthToken, where if you clear pin/fingerprint, you lose the reauthToken and have to sign in again.